pkg/cli/client_url.go

```diff
@@ -160,6 +155,74 @@ func (u urlParser) setInternal(v string, warn bool) error {
 		cliCtx.Insecure = insecure
 	}
 
+	makeStrictErr := func() error {
+		return fmt.Errorf("command %q only supports sslmode=disable or sslmode=verify-full", u.cmd.Name())
+	}
+
+	purl, err := updateURL(v,
+		u.sslStrict,
+		makeStrictErr,
+		usernameFlag,
+		foundUsername,
+		foundHostname,
+		foundPort,
+		foundDatabase,
+		insecureFlag,
+		insecureOverride,
+		certsDirFlag,
+		foundCertsDir,
+	)
+
+	if err != nil {
+		// This function is called by pflag.(*FlagSet).Set() and that code
+		// does not know how to use errors.Wrap properly. Instead, it
+		// reformats the error as a string, which loses any validation
+		// details.
+		// We make-do here by injecting the details as part of
+		// the error message.
+		// TODO(knz): Fix the upstream pflag and get rid of this
+		// horrendous logic.
+		msg := err.Error()
+		if details := errors.FlattenDetails(err); details != "" {
+			msg += "\n" + details
+		}
+		return fmt.Errorf("%s", msg)
+	}
+
+	// Store the parsed URL for later.
+	cliCtx.sqlConnURL = purl
+	return err
+}
+
+// updateURL is a helper that processes a connection URL passed on the
+// command line, integrates any defaults set by earlier command-line arguments,
+// then calls callbacks depending on which details were found in the new URL.
+//
+// sslStrict, when set to true, requires that the SSL file paths in
+// a URL clearly map to a single certificate directory and restricts
+// the set of supported SSL modes to just "disable" and "require".
+//
+// This is set for all non-SQL client commands, which only support
+// the insecure boolean and certs-dir with maximum SSL validation.
+func updateURL(
+	newURL string,
+	sslStrict bool,
+	makeStrictErr func() error,
```

Comment:
optional: how do you feel about these functions being an interface instead? 
could add unit tests too, if we wanted to. having to pass a lot of hooks in is easy to lose track of.

Reply:
good idea. done (in split PR)